Minor enhancements for teaching and learning about trees.

georgeflanagin opened the pull request:
The comments in the top of rb_tree.py cover the changes. 

I am an organizer of PyRVA.org, and I work at University of Richmond where I sometimes supply advice about Python programming.

georgeflanagin committed:
Minor enhancements for teaching and learning about trees.

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+of having it look somewhat more like the rest of the code at the
+university. The changes are only additions.
+
+ 1. The color of the nodes is now an Enum.
```
I'd prefer we keep the changes outside the source file itself. This is a very good description for the git commit message and PR description.

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
-RED = 'RED'
-NIL = 'NIL'
+import typing
+from   typing import *
```
nit: there are two extra spaces after from. `from   typing` -> `from typing`

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
-BLACK = 'BLACK'
-RED = 'RED'
-NIL = 'NIL'
+import typing
```
This import isn't used as far as I can tell

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+from   typing import *
+
+import enum
+import inspect
```
This import also isn't used as far as I can tell

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+__copyright__ = 'Copyright 2018, Stanislavkozlovski'
+__credits__ = None
+__version__ = '1.1'
+__maintainer__ = 'George Flanagin'
```
Does this mean you commit to maintaining the repo?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+__credits__ = None
+__version__ = '1.1'
+__maintainer__ = 'George Flanagin'
+__email__ = 'gflanagin@richmond.edu'
```
It's a bit unclear whose email this is

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+__version__ = '1.1'
+__maintainer__ = 'George Flanagin'
+__email__ = 'gflanagin@richmond.edu'
+__status__ = 'Production'
```
I vote we remove this line. Yes the code is polished and seemingly works, but I'd rather not advertise it as something that is "production-ready"

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+
+
+""" For the purpose of forward references. """
+class Node:
```
Very good idea with creating it for forward references. Can we change the comment to either be `# ...` or have it be a docstring (between class Node and pass)

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+    Nodes in any useful Tree.
+    """
+
+    __slots__ = [ 'value', 'color', 'parent', 'left', 'right' ]
```
nit: there is space in between brackets `[ ` ` ]`

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+    def __init__(self, value:Any, 
+        color:NODE_COLOR, 
+        parent: Node=None, 
+        left: Node =None, 
```
nit: there is spacing after Node here

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+    __slots__ = [ 'value', 'color', 'parent', 'left', 'right' ]
+
+    def __init__(self, value:Any, 
+        color:NODE_COLOR, 
```
nit: no space after semicolon in color and value arguments

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
 
-    def __eq__(self, other):
-        if self.color == NIL and self.color == other.color:
+    def __eq__(self, other:Node):
```
nit: space after colon

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+        """
+    
+        # Case 0: The argument is not a fellow Node.
+        if not isinstance(other, Node): return NotImplemented
```
Can we stick to PEP8 and have the return statement be on the bottom?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+            parents_are_same = ( self.parent.value == other.parent.value and 
+                        self.parent.color == other.parent.color )
+
+        return ( self.value == other.value and 
```
nit: space after parentheses

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+                        self.parent.color == other.parent.color )
+
+        return ( self.value == other.value and 
+                 self.color == other.color and 
```
I'm not certain whether this is in the PEP8 style guide but I recall seeing it be common Python style - please place the `and` in front of the condition in all cases rather than have it be trailing

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
-    def __init__(self):
+class RedBlackTree:
+    """
+    This is an implementation of the data structure described in the paper
```
Very good addition 👍

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+    def __init__(self) -> None:
         self.count = 0
         self.root = None
+        self.type = type(None)
```
Should this be `self.value_type`?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+        their __iter__ functions.
+        """ 
+
+        if not self.root: return []
```
Let's please follow PEP8 and have the return statement be on the next line

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
-    def add(self, value):
+
+    def iter_dump(self) -> str:
+        if not self.root: return [""]
```
Let's please follow PEP8 and have the return statement be on the next line. Also empty strings should use single quotes `''`

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
+        return "\n".join(r)
+
+
+    def add(self, value:Any) -> None:
```
nit: space after colon

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,287 @@
-            return  # value is in the tree
-        new_node = Node(value=value, color=RED, parent=parent, left=self.NIL_LEAF, right=self.NIL_LEAF)
+
+        if node_dir is None: # it is already here.
```
nit: I prefer the previous comment `value is in the tree`. Maybe have it be `value is already in the tree` ?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -84,142 +290,172 @@ def add(self, value):
 
-    def remove(self, value):
+
+    def remove(self, value:Any):
```
nit: space after colon

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -84,142 +290,172 @@ def add(self, value):
-    def contains(self, value) -> bool:
-        """ Returns a boolean indicating if the given value is present in the tree """
+
+    def contains(self, value:Any) -> bool:
```
nit: space after colon

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -84,142 +290,172 @@ def add(self, value):
         return bool(self.find_node(value))
 
-    def ceil(self, value) -> int or None:
+    def ceil(self, value:Any) -> Union[Node, None]:
```
nit: space after colon

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -84,142 +290,172 @@ def add(self, value):
         last_found_val = None if self.root.value < value else self.root.value
 
-        def find_ceil(node):
+        def find_ceil(node:None) -> Node:
```
Why should the node be of type `None`?
nit: space after colon.

stanislavkozlovski reviewed (changes requested):
Hi George,

These changes look very good. You have certainly improved the codebase.
There is some incongruence in style which I am not happy with. I think this project should follow the common PEP8 style guide which almost all python projects follow. It makes sense, since it is the official style guide and the most familiar to the biggest amount of people.

Do you mind changing the code such that it is PEP8 compliant?
If you lack the time, we could probably merge this into another branch where I'll review again but hold back stylistic comments. When merged, I could sort out the style myself.

Commit:
Minor additions for the purpose of teaching about trees.

 1. The color of the nodes is now an Enum.
 2. The Node class contains __slots__ to save space.
 3. For the purposes of teaching, I added a second iterator that
    prints the nodes of the tree with their connectivity.
 4. There is a very tiny demo at the end of the file.
 5. Also for the purposes of teaching about trees, I put in a type
    check to ensure all the Nodes in a tree are of the same
    value type.
 6. I added two one-line @classmethods to build red and black
    nodes. This makes it easier (for me) to see when the two
    types of nodes are being built.
 7. I added typing compatible with Python 3.6.
 8. I added traceback and inspect to assist with debugging,
    again primarily as a teaching aid.
 9. I added a __len__ operator to the tree to retun the number
    of Nodes, which conforms to standard practice.
10. I added a small number of boolean functions as an aid to
    teaching.

Commit:
Missed one change.

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,260 @@
-RED = 'RED'
-NIL = 'NIL'
+import sys
+import typing
```
Unused import

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,260 @@
+    def __len__(self) -> int:
+        return self.count
+
+    def __stack_trace(self) -> str:
```
This method can be static. What do you think if we make it a file-level function?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,260 @@
-    def __repr__(self):
+    """ Two @classmethods to build new nodes of a given color. """
+    @classmethod
+    def BlackNode(Node_cls, value: Any) -> Node:
```
The convention is to use `cls` as the name of the first variable in a class method

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -1,81 +1,260 @@
+        return Node(value, NODE_COLOR.BLACK, None, NIL_LEAF, NIL_LEAF)
+
+    @classmethod
+    def RedNode(Node_cls, value: Any, parent: Node) -> Node:
```
The convention is to use `cls` as the name of the first variable in a class method

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -84,142 +263,169 @@ def add(self, value):
+                s = ('The red child of a black node with 0 or 1 children' +
+                     ' cannot have children, otherwise the black height of' +
+                     ' the tree becomes invalid!\n')
+                raise(s + self.__stack_trace())
```
Does `raise(String)` even work? I locally tested with `raise("aa")` and it raises a `TypeError: exceptions must derive from BaseException`

stanislavkozlovski replied:
I think you should simply continue from the original code:
```python
raise Exception('The red child of a black node with 0 or 1 children'	                s = ('The red child of a black node with 0 or 1 children' +
                                ' cannot have children, otherwise the black height of the tree becomes invalid! {}'.format(self.__stack_trace()))
```

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -84,142 +263,169 @@ def add(self, value):
-            else:  # BLACK child
-                # 6 cases :o
+
+            else:  # NODE_COLOR.BLACK child
```
I guess this was an automatic change from your IDE/text editor. Maybe the comment should only be `black child` ?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -84,142 +263,169 @@ def add(self, value):
-    def _remove_leaf(self, leaf):
-        """ Simply removes a leaf node by making it's parent point to a NIL LEAF"""
+    def _remove_leaf(self, leaf: Node) -> None:
+        """ Simply removes a leaf node by making it's parent point to a NODE_COLOR.NIL LEAF"""
```
Ditto here, maybe only `nil leaf` ?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -230,16 +436,17 @@ def __case_1(self, node):
-            the sibling's children are BLACK or NIL
+            the parent is NODE_COLOR.BLACK
+            the sibling is NODE_COLOR.RED
+            the sibling's children are NODE_COLOR.BLACK or NODE_COLOR.NIL
```
Ditto

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -248,24 +455,25 @@ def __case_2(self, node):
                          50B    80B                                |20B|  50B
             (if the sibling's direction was left of it's parent, we would RIGHT ROTATE it)
-        Now the original node's parent is RED
+        Now the original node's parent is NODE_COLOR.RED
```
Ditto

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -248,24 +455,25 @@ def __case_2(self, node):
-            self.ROTATIONS[direction](node=None, parent=sibling, grandfather=parent)
-            parent.color = RED
-            sibling.color = BLACK
+        if silbing.is_red and parent.is_black and sibling.left.is_red and sibling.right.is_not_red:
```
This changes behavior. The sibling should not be red?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -248,24 +455,25 @@ def __case_2(self, node):
-            the sibling is BLACK
-            the sibling's children are BLACK
+            the parent is NODE_COLOR.BLACK
+            the sibling is NODE_COLOR.BLACK
```
ditto

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -280,16 +488,15 @@ def __case_3(self, node):
         sibling, _ = self._get_sibling(node)
-        if (sibling.color == BLACK and parent.color == BLACK
-           and sibling.left.color != RED and sibling.right.color != RED):
+        if sibling.is_black and parent.is_black and sibling.left.is_not_red and sibling.right.is_not_red:
```
I think this becomes too long of a line, mind breaking it up like before?

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -301,17 +508,17 @@ def __case_4(self, node):
         """
         Case 5 is a rotation that changes the circumstances so that we can do a case 6
-        If the closer node is red and the outer BLACK or NIL, we do a left/right rotation, depending on the orientation
+        If the closer node is red and the outer NODE_COLOR.BLACK or NODE_COLOR.NIL, we do a left/right rotation, depending on the orientation
```
Ditto

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -325,21 +532,23 @@ def __case_5(self, node):
         Case 6 requires
-            SIBLING to be BLACK
-            OUTER NODE to be RED
+            SIBLING to be NODE_COLOR.BLACK
```
Dito

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -348,7 +557,7 @@ def __case_6(self, node):
       SIBLING--> 35B      |80B| <-                             30R       50R
                 /   \      /                                  /   \     /   \
-             30R    37B  70R   Outer node is RED            20B   34B 37B    80B
+             30R    37B  70R   Outer node is NODE_COLOR.RED            20B   34B 37B    80B
```
Ditto

stanislavkozlovski commented on rb_tree.py:
```diff
@@ -360,59 +569,78 @@ def __case_6(self, node):
-           or (node.color != RED or parent.color != RED)):  # no need to rebalance
+
+        if (parent is None or        # what the fuck? (should not happen)
+            parent.parent is None or  # parent is the root
```
Can we move the `or`s in the beginning?

stanislavkozlovski reviewed (changes requested):
Thanks for the changes. I've left some minor comments on the current implementation.

There is one bigger problem here, though - the tests in `rb_tree_tests.py` have not been updated and as a result won't even "compile".
